### PR TITLE
added the add bool type argument in model for autocomplete

### PR DESCRIPTION
**Pull request by KrishnaSindhur**
#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/14739

### CODE REVIEW

**Comment by hanzei** [comment]
Thanks for the PR @KrishnaSindhur, I've queued it for reivew. Heads up that CI fails with:
```
Failed
    TestAutocompleteDataJSON: command_autocomplete_test.go:83: 
```

**Comment by KrishnaSindhur** [comment]
ya i am on it

**Comment by KrishnaSindhur** [comment]
@hanzei what is the EETests failed means?

**Comment by hanzei** [comment]
@KrishnaSindhur Your code was run alongside the Enterprise closed source code and the tests there. These tests failed due to flakiness. I did restart the tests. Hopefully the will run fine.

**Comment by KrishnaSindhur** [comment]
@iomodo  ohh i had that doubt sure then i can check from api also rite its end to end then

File: app/command_autocomplete.go [hunk 1]
```diff
@@ -251,3 +259,23 @@ func parseListItems(items []model.AutocompleteListItem, parsed, toBeParsed strin
 	}
 	return true, parsed + toBeParsed, "", suggestions
 }
+
+func parseBoolArgument(arg *model.AutocompleteArg, parsed, toBeParsed string) (found bool, alreadyParsed string, yetToBeParsed string, suggestion model.AutocompleteSuggestion) {
+	in := strings.TrimPrefix(toBeParsed, " ")
+	a := arg.Data.(*model.AutocompleteBoolArg)
+	if in == "" { //The user has not started typing the argument.
+		return true, parsed + toBeParsed, "", model.AutocompleteSuggestion{Complete: parsed + toBeParsed, Suggestion: "", Hint: a.Hint, Description: arg.HelpText}
```
Comment (iomodo) [inline app/command_autocomplete.go:267]: We should probably return `true` and `false` as suggestions here

Comment (KrishnaSindhur) [reply app/command_autocomplete.go:267]: yes sure i will update that

File: app/command_autocomplete.go [hunk 2]
```diff
@@ -174,19 +174,35 @@ func (a *App) parseArgument(commandArgs *model.CommandArgs, arg *model.Autocompl
...
+		parsed = changedParsed
+		toBeParsed = changedToBeParsed
 	}
 
 	return false, parsed, toBeParsed, suggestions
 }
 
 func parseNamedArgument(arg *model.AutocompleteArg, parsed, toBeParsed string) (found bool, alreadyParsed string, yetToBeParsed string, suggestion model.AutocompleteSuggestion) {
 	in := strings.TrimPrefix(toBeParsed, " ")
+	if len(arg.Name) == 0 {
```
Comment (iomodo) [inline app/command_autocomplete.go:192]: Named argument can be can be a text argument or list argument as well as boolean. With this implementation every named argument will be autocompleted with `true` or `false`. What you have to do is check if we have bool named argument on our hand.

**Review by iomodo — CHANGES_REQUESTED** [review]
Great work @KrishnaSindhur! See one suggestion bellow.
Also it will be great to have the some implementation of the bool arguments in the plugin for testing purposes. It will be great if you could add couple of examples in the [demo plugin](https://www.google.com/search?client=safari&rls=en&q=github+demo+plugin+mattermost&ie=UTF-8&oe=UTF-8)

File: app/command_autocomplete.go (hunk 2 above)
Comment (KrishnaSindhur) [reply app/command_autocomplete.go:192]: sure i will have check named argument is bool type then only i will add the autocomplete true or false

**Comment by KrishnaSindhur** [comment]
> demo plugin

sure i will add that

**Comment by jfrerich** [comment]
@levb, @iomodo Ping for review

**Review by levb — APPROVED** [review]

File: app/command_autocomplete_test.go [hunk 3]
```diff
@@ -132,71 +132,166 @@ func TestParseInputTextArgument(t *testing.T) {
...
+	assert.Equal(t, "false ", parsed)
+	assert.Equal(t, "test_command", toBeParsed)
+
+	found, parsed, toBeParsed, _ = parseBoolArgument(argument, "", "abc ")
+	assert.False(t, found)
+	assert.Equal(t, "abc ", parsed)
+	assert.Equal(t, "", toBeParsed)
+}
+
 func TestSuggestions(t *testing.T) {
```
Comment (iomodo) [inline app/command_autocomplete_test.go:295]: Could we add some unit tests here? In the `TestCommandWithOptionalArgs` function as well. Let's test bool arguments in the complex environment with different type of args, including optional args.

**Review by iomodo — APPROVED** [review]
Looks good @KrishnaSindhur one minor ask below.

File: app/command_autocomplete_test.go (hunk 3 above)
Comment (KrishnaSindhur) [reply app/command_autocomplete_test.go:295]: sure i will add that

**Comment by levb** [comment]
@KrishnaSindhur Not sure what happened here, but I can no longer easily review this, can you please re-base/merge and push a mergeable branch?

**Comment by KrishnaSindhur** [comment]
> @KrishnaSindhur Not sure what happened here, but I can no longer easily review this, can you please re-base/merge and push a mergeable branch?

i think i can't make any changes to this branch

**Comment by levb** [comment]
@KrishnaSindhur if you make a clean version by starting with `master`, then cherry-picking your PRs on top, or even squash them... either way, once you have a clean, mergeable branch with your changes we can force-push it into here to keep the PR and its comments.

**Comment by KrishnaSindhur** [comment]
> @KrishnaSindhur if you make a clean version by starting with `master`, then cherry-picking your PRs on top, or even squash them... either way, once you have a clean, mergeable branch with your changes we can force-push it into here to keep the PR and its comments.
added all the the changes along with new tests related to story to here [https://github.com/mattermost/mattermost-server/pull/15923](url)

**Comment by hanzei** [comment]
Replaced by https://github.com/mattermost/mattermost-server/pull/15923

**Pull request closed without merging**